R/functions-Spectrum.R

```diff
@@ -568,3 +587,462 @@ validSpectrum <- function(object) {
+    mzs_lens <- base::lengths(mzs)
+    mzs <- unlist(mzs, use.names = FALSE)
+    mz_order <- base::order(mzs)
+    mzs <- mzs[mz_order]
+    mz_groups <- .group_mz_values(mzs, mzd = mzd)
+    if (length(unique(mz_groups)) < length(x[[main]]@mz))
+        stop("Got less m/z groups than m/z values in the original spectrum. ",
+             "Most likely the data is not profile-mode LCMS data.")
+    ## Want to keep only those groups with a m/z from the main spectrum.
+    keep <- unlist(lapply(
+        split(
+            rep(1:length(mzs_lens), mzs_lens)[mz_order] == main,
+            mz_groups),
+        function(z) {
+            rep(any(z), length(z))
+        }), use.names = FALSE)
+    ## Keep only values for which a m/z in main is present.
+    mz_groups <- mz_groups[keep]
+    mzs <- mzs[keep]
+    ints <- unlist(base::lapply(x, function(z) z@intensity))[mz_order][keep]
+    ## Create result.
+    new_sp <- x[[main]]
+    ## Support also weighted.mean:
+    if (is.character(mzFun) && mzFun == "weighted.mean") {
+        intsp <- split(ints, mz_groups)
+        new_sp@mz <- base::mapply(split(mzs, mz_groups), intsp,
+                                  FUN = function(mz_vals, w)
+                                      stats::weighted.mean(mz_vals, w + 1,
+                                                           na.rm = TRUE),
+                                  USE.NAMES = FALSE, SIMPLIFY = TRUE)
+        new_sp@intensity <- unlist(base::lapply(intsp, intensityFun),
+                                   use.names = FALSE)
+    } else {
+        new_sp@mz <- unlist(base::lapply(split(mzs, mz_groups), mzFun),
+                            use.names = FALSE)
+        new_sp@intensity <- unlist(base::lapply(split(ints, mz_groups),
+                                                intensityFun),
+                                   use.names = FALSE)
+    }
+    if (is.unsorted(new_sp@mz))
+        stop("m/z values of combined spectrum are not ordered")
+    new_sp@peaksCount <- length(new_sp@mz)
+    new_sp
+}
+
+#' Group a sorted `numeric` of m/z values from consecutive scans by ion assuming
+#' that the variation between m/z values for the same ion in consecutive scan
+#' is much lower than the difference between m/z values within one scan.
+#'
+#' @param x `numeric` with ordered and combined m/z values from consecutive
+#'     scans.
+#'
+#' @param mzd `numeric(1)` with the m/z difference below which m/z values are
+#'     grouped together. If not provided the `.estimate_mz_scattering` function
+#'     is used to estimate it.
+#' 
+#' @return `integer` of same length than `x` grouping m/z values.
+#'
+#' @noRd
+.group_mz_values <- function(x, mzd) {
+    mzdiff <- diff(x)
+    if (missing(mzd))
+        mzd <- .estimate_mz_scattering(mzdiff, TRUE)
+    ## Create a vector grouping values with difference in mz values being
+    ## smaller than mzd
+    nvals <- diff(c(0, which(!(mzdiff < mzd)), length(x)))
+    rep(1:length(nvals), nvals)
```

Comment:
I would suggest to use `cumsum(c(0L, mzdiff >= mzd)) + 1L` instead of `nvals <- diff(c(0, which(!(mzdiff < mzd)), length(x))); rep(1:length(nvals), nvals)` (for my test cases this is nearly 3 times faster).

Reply:
Thanks!